pure-html/pure_html.ml

```diff
@@ -51,41 +59,101 @@ let xml_mode xml name =
   | true, _ | _, ("math" | "svg") -> true
   | _ -> false
 
-(* Loosely based on https://www.w3.org/TR/DOM-Parsing/ *)
-let rec write_tag ~xml p = function
+module Indent_level = struct
+  type t = int option
+
+  let next (t : t) =
+    match t with
+    | None -> None
+    | Some n -> Some (n + 1)
+
+  (* The following functions return string that should be inserted before or
+    after the open or closing tag with indent level t. Some of them also need to
+    know if the next tag is going to be indented one more level. *)
+
+  let open_prefix (t : t) =
+    match t with
+    | None -> ""
+    | Some n -> String.make (n * 2) ' '
+
+  let open_suffix (t : t) ~indent_next =
+    match t, indent_next with
+    | Some _, true -> "\n"
+    | _, _ -> ""
+
+  let close_prefix (t : t) ~indent_next =
+    match t, indent_next with
+    | Some n, true -> String.make (n * 2) ' '
+    | _, _ -> ""
+
+  let close_suffix (t : t) =
+    match t with
+    | None -> ""
+    | Some _ -> "\n"
+end
+
+(* Loosely based on https://www.w3.org/TR/DOM-Parsing/. Pretty prints using two
+  spaces for indentation. On high level, the algorithm indents the children of a
+  tag if they do not contain a txt node. No newline is inserted if there are no
+  children nodes. *)
+let rec write_tag ~indent_level ~xml p node =
+  let should_indent_next children =
```

Comment:
This function doesn't need to be nested; we can lift it up.

Reply:
Updated.